hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java

```diff
@@ -1782,6 +1782,29 @@ public BalanceResponse balance() throws IOException {
     return balance(BalanceRequest.defaultInstance());
   }
 
+  /**
+   *  Trigger a normal balance, see {@link HMaster#balance()} . If the balance is not executed
+   *  this time, the metrics related to the balance will be updated.
+   *
+   *  When balance is running, related metrics will be updated at the same time. But if some
+   *  checking logic failed and cause the balancer exit early, we lost the chance to update
+   *  balancer metrics. This will lead to user missing the latest balancer info.
+   * */
+  public BalanceResponse balanceOrUpdateMetrics() throws IOException{
+    BalanceResponse response = balance(BalanceRequest.defaultInstance());
+    if (!response.isBalancerRan()) {
+      Map<TableName, Map<ServerName, List<RegionInfo>>> assignments =
```

Comment:
i think we should wrap this in a `synchronized (this.balancer)` now that it's in HMaster.java, to resolve any conflicts around the balancer's mutable fields that updateBalancerLoadInfo touches.

This does theoretically open us up to a race condition where another thread was waiting on the synchronization in `balance()`, and then when the balancer finished the other thread got into that synchronized block. At which point we'd then wait here. We could get around that by wrapping this entire method in a synchronized block, but honestly I might just be fine allowing the off chance of the race condition since the side-effect of that is relatively trivial.

Reply:
Got it! I now think it would be better to put them into a synchronized block.